server/lib/subscriptions.ts

```diff
@@ -21,12 +21,14 @@ const getIsSubscriptionManagedExternally = pm => {
 };
 
 /**
- * When the contribution gets updated, we need to update the next charge date as well
+ * When the contribution gets updated, we need to update the next charge date as well.
+ *
+ * If reactivating the contribution on the same interval (month/year), keep the previous "next charge date". Otherwise,
+ * set the next charge date to the beginning of the next interval.
  */
-const getNextChargeDateForUpdateContribution = (baseNextChargeDate, newInterval) => {
+const getNextChargeDateForUpdateContribution = (baseNextChargeDate: Date, newInterval: Order['interval']) => {
   const previousNextChargeDate = moment(baseNextChargeDate);
-  if (previousNextChargeDate.isBefore(moment())) {
-    // If the contribution was pending, keep it in the past
+  if (previousNextChargeDate.isSame(moment(), newInterval)) {
```

Comment:
The logic change here will impact all types of subscription updates: when updating the payment method of a subscription, we will only charge the "previousNextChargeDate" if it's in the same interval (current month or year).